File: src/services/inbox.ts
/**
 * Inbox service (§12): the ONLY human interrupt channel. Agents send text
 * messages or multiple-choice questions; the human replies in the UI and the
 * waiting session resumes with the answer.
 */
import { randomUUID } from "node:crypto";
import { eq } from "drizzle-orm";
import { inboxMessages } from "../db/schema.js";
import type { DB } from "../db/client.js";
import type { InboxMessage, InboxChoice } from "../domain/types.js";
import { HttpError } from "../api/errors.js";

export interface InboxSendInput {
  from: "agent" | "human";
  agentId?: string | null;
  sessionId?: string | null;
  taskId?: string | null;
  goalId?: string | null;
  kind?: "text" | "multiple-choice";
  body: string;
  choices?: string[];
}

export class InboxService {
  constructor(private db: DB) {}

  async send(input: InboxSendInput): Promise<InboxMessage> {
    const row: InboxMessage = {
      id: randomUUID(),
      from: input.from,
      agentId: input.agentId ?? null,
      sessionId: input.sessionId ?? null,
      taskId: input.taskId ?? null,
      goalId: input.goalId ?? null,
      kind: input.kind ?? "text",
      body: input.body,
      choices: (input.choices ?? []).map((c, i) => ({ id: `c${i}`, label: c }) satisfies InboxChoice),
      selectedChoiceId: null,
      status: "open",
      createdAt: new Date().toISOString(),
    };
    await this.db.insert(inboxMessages).values(row).run();
    return row;
  }

  async list(): Promise<InboxMessage[]> {
    const rows = await this.db.select().from(inboxMessages).all();
    return rows.sort((a, b) => a.createdAt.localeCompare(b.createdAt));
  }

  async get(id: string): Promise<InboxMessage | null> {
    const row = await this.db.select().from(inboxMessages).where(eq(inboxMessages.id, id)).get();
    return row ?? null;
  }

  async openForSession(sessionId: string): Promise<InboxMessage[]> {
    const rows = await this.db.select().from(inboxMessages).where(eq(inboxMessages.sessionId, sessionId)).all();
    return rows.sort((a, b) => a.createdAt.localeCompare(b.createdAt));
  }

  /**
   * Human reply. Records the reply (choice or text) and returns the message
   * the reply answers. The session service resumes the waiting session.
   */
  async reply(
    messageId: string,
    input: { body?: string; selectedChoiceId?: string },
  ): Promise<InboxMessage> {
    const msg = await this.get(messageId);
    if (!msg) throw new HttpError(404, "message not found");
    if (msg.status !== "open") throw new HttpError(409, "message already answered");
    if (msg.kind === "multiple-choice" && !input.selectedChoiceId) {
      throw new HttpError(400, "multiple-choice message requires selectedChoiceId");
    }
    if (input.selectedChoiceId && !msg.choices.some((c) => c.id === input.selectedChoiceId)) {
      throw new HttpError(400, "invalid choice id");
    }
    const answer: InboxMessage = {
      id: randomUUID(),
      from: "human",
      agentId: msg.agentId,
      sessionId: msg.sessionId,
      taskId: msg.taskId,
      goalId: msg.goalId,
      kind: "text",
      body: input.body ?? `(selected: ${msg.choices.find((c) => c.id === input.selectedChoiceId)?.label ?? input.selectedChoiceId})`,
      choices: [],
      selectedChoiceId: null,
      status: "closed",
      createdAt: new Date().toISOString(),
    };
    await this.db.insert(inboxMessages).values(answer).run();
    const updated = { ...msg, status: "answered" as const, selectedChoiceId: input.selectedChoiceId ?? null };
    await this.db.update(inboxMessages).set(updated).where(eq(inboxMessages.id, msg.id)).run();
    return updated;
  }

  async close(messageId: string): Promise<void> {
    await this.db.update(inboxMessages).set({ status: "closed" }).where(eq(inboxMessages.id, messageId)).run();
  }
}

File: agentos/services/inbox.py
from __future__ import annotations
"""Inbox service. Port of src/services/inbox.ts."""
import uuid
from datetime import datetime, timezone
from typing import Optional

from sqlalchemy import select, update

from agentos.db.client import get_session
from agentos.db.models import InboxMessage as InboxMessageRow


class HttpError(Exception):
    def __init__(self, status: int, message: str) -> None:
        super().__init__(message)
        self.status = status
        self.message = message


class InboxService:
    async def send(self, *, from_: str, agent_id: str | None = None,
                   session_id: str | None = None, task_id: str | None = None,
                   goal_id: str | None = None, kind: str = "text",
                   body: str, choices: list[str] | None = None) -> dict:
        row = InboxMessageRow(
            id=str(uuid.uuid4()),
            from_=from_,
            agent_id=agent_id,
            session_id=session_id,
            task_id=task_id,
            goal_id=goal_id,
            kind=kind,
            body=body,
            choices=[{"id": f"c{i}", "label": c} for i, c in enumerate(choices or [])],
            selected_choice_id=None,
            status="open",
            created_at=datetime.now(timezone.utc).isoformat(),
        )
        async with get_session() as db:
            db.add(row)
            await db.commit()
        return _row_to_dict(row)

    async def list(self) -> list[dict]:
        async with get_session() as db:
            result = await db.execute(select(InboxMessageRow).order_by(InboxMessageRow.created_at))
            return [_row_to_dict(r) for r in result.scalars().all()]

    async def get(self, msg_id: str) -> dict | None:
        async with get_session() as db:
            result = await db.execute(select(InboxMessageRow).where(InboxMessageRow.id == msg_id))
            row = result.scalar_one_or_none()
        return _row_to_dict(row) if row else None

    async def open_for_session(self, session_id: str) -> list[dict]:
        async with get_session() as db:
            result = await db.execute(
                select(InboxMessageRow)
                .where(InboxMessageRow.session_id == session_id)
                .order_by(InboxMessageRow.created_at)
            )
            return [_row_to_dict(r) for r in result.scalars().all()]

    async def reply(self, message_id: str, body: str | None = None,
                    selected_choice_id: str | None = None) -> dict:
        async with get_session() as db:
            result = await db.execute(select(InboxMessageRow).where(InboxMessageRow.id == message_id))
            msg = result.scalar_one_or_none()
            if not msg:
                raise HttpError(404, "message not found")
            if msg.status != "open":
                raise HttpError(409, "message already answered")
            if msg.kind == "multiple-choice" and not selected_choice_id:
                raise HttpError(400, "multiple-choice message requires selectedChoiceId")
            if selected_choice_id and not any(c["id"] == selected_choice_id for c in (msg.choices or [])):
                raise HttpError(400, "invalid choice id")

            # Record human answer
            answer_row = InboxMessageRow(
                id=str(uuid.uuid4()),
                from_="human",
                agent_id=msg.agent_id,
                session_id=msg.session_id,
                task_id=msg.task_id,
                goal_id=msg.goal_id,
                kind="text",
                body=body or f"(selected: {next((c['label'] for c in (msg.choices or []) if c['id'] == selected_choice_id), selected_choice_id)})",
                choices=[],
                selected_choice_id=None,
                status="closed",
                created_at=datetime.now(timezone.utc).isoformat(),
            )
            db.add(answer_row)
            msg.status = "answered"
            msg.selected_choice_id = selected_choice_id
            await db.commit()
            return _row_to_dict(msg)

    async def close(self, message_id: str) -> None:
        async with get_session() as db:
            result = await db.execute(select(InboxMessageRow).where(InboxMessageRow.id == message_id))
            row = result.scalar_one_or_none()
            if row:
                row.status = "closed"
                await db.commit()


def _row_to_dict(r: InboxMessageRow) -> dict:
    return {
        "id": r.id, "from": r.from_, "agentId": r.agent_id,
        "sessionId": r.session_id, "taskId": r.task_id, "goalId": r.goal_id,
        "kind": r.kind, "body": r.body, "choices": r.choices or [],
        "selectedChoiceId": r.selected_choice_id, "status": r.status,
        "createdAt": r.created_at,
    }
